modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -0,0 +1,400 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CrystalPlasticityStressUpdateBase.h"
+
+#include "libmesh/utility.h"
+#include "Conversion.h"
+#include "MooseException.h"
+
+InputParameters
+CrystalPlasticityStressUpdateBase::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addParam<std::string>(
+      "base_name",
+      "Optional parameter that allows the user to define multiple crystal plasticity mechanisms");
+  params.addClassDescription(
+      "Crystal Plasticity base class: handles the Newton iteration over the stress residual and "
+      "calculates the Jacobian based on constitutive laws provided by inheriting classes");
+
+  // The return stress increment classes are intended to be iterative materials, so must set compute
+  // = false for all inheriting classes
+  params.set<bool>("compute") = false;
+  params.suppressParameter<bool>("compute");
+
+  params.addRequiredParam<unsigned int>(
+      "number_slip_systems",
+      "The total number of possible active slip systems for the crystalline material");
+  params.addRequiredParam<FileName>("slip_sys_file_name",
+                                    "Name of the file containing the slip systems");
+  params.addParam<Real>("number_cross_slip_directions",
+                        0,
+                        "Quanity of unique slip directions, used to determine cross slip familes");
+  params.addParam<Real>("number_cross_slip_planes",
+                        0,
+                        "Quanity of slip planes belonging to a single cross slip direction; used "
+                        "to determine cross slip families");
+  params.addParam<Real>(
+      "slip_increment_tolerance",
+      2e-2,
+      "Maximum allowable slip in an increment for each individual constitutive model");
+  params.addParam<Real>(
+      "stol", 1e-2, "Constitutive internal state variable relative change tolerance");
+  params.addParam<Real>("resistance_tol",
+                        1.0e-2,
+                        "Constitutive slip system resistance relative residual tolerance for each "
+                        "individual constitutive model");
+  params.addParam<Real>("zero_tol",
+                        1e-12,
+                        "Tolerance for residual check when variable value is zero for each "
+                        "individual constitutive model");
```

Comment:
These are put in the `CrystalPlasticityStressUpdateBase` class to enable different tolerances for different constitutive models, which may help the convergence when users are using a combination of many constitutive models, one or some of which are highly nonlinear thus difficult to converge. 

Putting the tolerances in the `ComputeMultipleCrystalPlasticityStress` class will make it easier for users to make decisions, but may not handle the case when different tolerances are necessary.

Suggestion here? @bwspenc @dschwen

Reply:
seems useful to have it in the update material (here) with a sensible default